File: lib/external-gtfs/gtfs-time.js
'use strict';
/**
 * Heures et dates GTFS. Aucun arrondi, aucune heure fabriquée : une heure est
 * soit lue telle quelle dans stop_times.txt, soit absente.
 *
 * Fuseau de service : Africa/Dakar = UTC+0 toute l'année (pas d'heure d'été).
 */

const TIME_RE = /^\s*(\d{1,2}):(\d{2})(?::(\d{2}))?\s*$/;

/** 'HH:MM[:SS]' (HH peut dépasser 24 en GTFS) → secondes depuis minuit, ou null. */
function parseTime(value) {
  if (value === null || value === undefined) return null;
  const m = TIME_RE.exec(String(value));
  if (!m) return null;
  const h = Number(m[1]);
  const min = Number(m[2]);
  const s = m[3] === undefined ? 0 : Number(m[3]);
  if (min > 59 || s > 59) return null;
  return h * 3600 + min * 60 + s;
}

const pad2 = (n) => String(n).padStart(2, '0');

/** secondes → 'HH:MM:SS' (HH non réduit modulo 24, conforme GTFS). */
function formatTime(seconds) {
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = seconds % 60;
  return `${pad2(h)}:${pad2(m)}:${pad2(s)}`;
}

/** secondes → 'HH:MM' (affichage). */
function formatHHMM(seconds) {
  return `${pad2(Math.floor(seconds / 3600))}:${pad2(Math.floor((seconds % 3600) / 60))}`;
}

/** 'YYYY-MM-DD' | 'YYYYMMDD' → 'YYYYMMDD' ; lève si invalide. */
function normalizeServiceDate(value) {
  const s = String(value === undefined || value === null ? '' : value).trim();
  let m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(s);
  if (!m) m = /^(\d{4})(\d{2})(\d{2})$/.exec(s);
  if (!m) throw new Error(`Date de service invalide : « ${value} »`);
  const y = Number(m[1]);
  const mo = Number(m[2]);
  const d = Number(m[3]);
  const dt = new Date(Date.UTC(y, mo - 1, d));
  if (dt.getUTCFullYear() !== y || dt.getUTCMonth() !== mo - 1 || dt.getUTCDate() !== d) {
    throw new Error(`Date de service inexistante : « ${value} »`);
  }
  return `${m[1]}${m[2]}${m[3]}`;
}

/** 'YYYYMMDD' → 'YYYY-MM-DD'. */
function toIsoDate(yyyymmdd) {
  const s = normalizeServiceDate(yyyymmdd);
  return `${s.slice(0, 4)}-${s.slice(4, 6)}-${s.slice(6, 8)}`;
}

const WEEKDAYS = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];

/** Jour de semaine GTFS ('monday'…'sunday') d'une date 'YYYYMMDD'. */
function weekdayOf(yyyymmdd) {
  const s = normalizeServiceDate(yyyymmdd);
  const dt = new Date(Date.UTC(Number(s.slice(0, 4)), Number(s.slice(4, 6)) - 1, Number(s.slice(6, 8))));
  return WEEKDAYS[dt.getUTCDay()];
}

/** Date 'YYYYMMDD' + n jours. */
function addDays(yyyymmdd, n) {
  const s = normalizeServiceDate(yyyymmdd);
  const dt = new Date(Date.UTC(Number(s.slice(0, 4)), Number(s.slice(4, 6)) - 1, Number(s.slice(6, 8)) + n));
  return `${dt.getUTCFullYear()}${pad2(dt.getUTCMonth() + 1)}${pad2(dt.getUTCDate())}`;
}

/** Comparaison lexicale de deux dates 'YYYYMMDD' (négatif, 0, positif). */
function compareDates(a, b) {
  const x = normalizeServiceDate(a);
  const y = normalizeServiceDate(b);
  return x < y ? -1 : x > y ? 1 : 0;
}

/**
 * Horloge de Dakar (UTC+0) pour un instant donné : jour de service et heure.
 * @param {Date} instant
 */
function dakarClock(instant) {
  const d = instant instanceof Date ? instant : new Date(instant);
  if (Number.isNaN(d.getTime())) throw new Error('Instant invalide');
  const seconds = d.getUTCHours() * 3600 + d.getUTCMinutes() * 60 + d.getUTCSeconds();
  return {
    serviceDate: `${d.getUTCFullYear()}${pad2(d.getUTCMonth() + 1)}${pad2(d.getUTCDate())}`,
    isoDate: `${d.getUTCFullYear()}-${pad2(d.getUTCMonth() + 1)}-${pad2(d.getUTCDate())}`,
    seconds,
    time: formatTime(seconds),
  };
}

module.exports = {
  parseTime, formatTime, formatHHMM, normalizeServiceDate, toIsoDate, weekdayOf, addDays, compareDates, dakarClock, WEEKDAYS,
};

File: flutter-src/lib/services/external_gtfs/gtfs_time.dart
// Heures et dates GTFS — miroir Dart de lib/external-gtfs/gtfs-time.js.
// Aucun arrondi, aucune heure fabriquée : une heure est soit lue telle quelle
// dans stop_times.txt, soit absente. Fuseau de service Africa/Dakar = UTC+0.

final RegExp _timeRe = RegExp(r'^\s*(\d{1,2}):(\d{2})(?::(\d{2}))?\s*$');
final RegExp _isoDateRe = RegExp(r'^(\d{4})-(\d{2})-(\d{2})$');
final RegExp _compactDateRe = RegExp(r'^(\d{4})(\d{2})(\d{2})$');

const List<String> gtfsWeekdays = <String>[
  'monday',
  'tuesday',
  'wednesday',
  'thursday',
  'friday',
  'saturday',
  'sunday',
];

/// 'HH:MM[:SS]' (HH peut dépasser 24 en GTFS) → secondes depuis minuit, ou null.
int? parseGtfsTime(String? value) {
  if (value == null) return null;
  final RegExpMatch? m = _timeRe.firstMatch(value);
  if (m == null) return null;
  final int h = int.parse(m.group(1)!);
  final int min = int.parse(m.group(2)!);
  final int s = m.group(3) == null ? 0 : int.parse(m.group(3)!);
  if (min > 59 || s > 59) return null;
  return h * 3600 + min * 60 + s;
}

String _pad2(int n) => n.toString().padLeft(2, '0');

/// secondes → 'HH:MM:SS' (HH non réduit modulo 24, conforme GTFS).
String formatGtfsTime(int seconds) {
  final int h = seconds ~/ 3600;
  final int m = (seconds % 3600) ~/ 60;
  final int s = seconds % 60;
  return '${_pad2(h)}:${_pad2(m)}:${_pad2(s)}';
}

/// secondes → 'HH:MM' (affichage).
String formatHHMM(int seconds) =>
    '${_pad2(seconds ~/ 3600)}:${_pad2((seconds % 3600) ~/ 60)}';

/// 'YYYY-MM-DD' | 'YYYYMMDD' → 'YYYYMMDD' ; lève [FormatException] si invalide.
String normalizeServiceDate(String value) {
  final String s = value.trim();
  RegExpMatch? m = _isoDateRe.firstMatch(s);
  m ??= _compactDateRe.firstMatch(s);
  if (m == null) throw FormatException('Date de service invalide : « $value »');
  final int y = int.parse(m.group(1)!);
  final int mo = int.parse(m.group(2)!);
  final int d = int.parse(m.group(3)!);
  final DateTime dt = DateTime.utc(y, mo, d);
  if (dt.year != y || dt.month != mo || dt.day != d) {
    throw FormatException('Date de service inexistante : « $value »');
  }
  return '${y.toString().padLeft(4, '0')}${_pad2(mo)}${_pad2(d)}';
}

/// 'YYYYMMDD' → 'YYYY-MM-DD'.
String toIsoDate(String value) {
  final String d = normalizeServiceDate(value);
  return '${d.substring(0, 4)}-${d.substring(4, 6)}-${d.substring(6, 8)}';
}

DateTime _utcOf(String yyyymmdd) {
  final String d = normalizeServiceDate(yyyymmdd);
  return DateTime.utc(
    int.parse(d.substring(0, 4)),
    int.parse(d.substring(4, 6)),
    int.parse(d.substring(6, 8)),
  );
}

/// Jour de semaine GTFS ('monday' … 'sunday') d'une date de service.
String weekdayOf(String yyyymmdd) => gtfsWeekdays[_utcOf(yyyymmdd).weekday - 1];

/// Date de service décalée de [n] jours.
String addDays(String yyyymmdd, int n) {
  final DateTime dt = _utcOf(yyyymmdd).add(Duration(days: n));
  return '${dt.year.toString().padLeft(4, '0')}${_pad2(dt.month)}${_pad2(dt.day)}';
}

/// Horloge de service (Dakar = UTC+0 toute l'année).
class DakarClock {
  const DakarClock(this.serviceDate, this.seconds);

  final String serviceDate;
  final int seconds;

  String get isoDate => toIsoDate(serviceDate);
  String get time => formatGtfsTime(seconds);
}

DakarClock dakarClock(DateTime instant) {
  final DateTime u = instant.toUtc();
  final String day =
      '${u.year.toString().padLeft(4, '0')}${_pad2(u.month)}${_pad2(u.day)}';
  return DakarClock(day, u.hour * 3600 + u.minute * 60 + u.second);
}
